### PR TITLE
[Run Timeline] Improve rendering performance + Don't wait for future ticks before rendering

### PR DESCRIPTION
**Pull request by salazarm**
## Summary & Motivation

- Don't make the Run Timeline wait for FutureTicks to resolve before considering it complete because future ticks pulls the workspace snapshot and is really slow while the other queries tends to be fast.
- ~75% improvement in rendering performance by fixing needless re-rendering, N^2 isAdHoc job logic, and excessive spread usage

## How I Tested These Changes

Jest tests.
Loaded the Run timeline on various organizations testing both uncached and cached loads and pagination forwards/backwards.

Performance comparison for a cached load of the RunTimeline for a particularly large org 

Before :1,770.62ms
Screenshot 2024-06-10 at 9 09 21 AM

After: 448ms
Screenshot 2024-06-10 at 9 07 08 AM

### CODE REVIEW

File: js_modules/dagster-ui/packages/ui-core/src/runs/RunTimeline.tsx [hunk 1]
```diff
@@ -93,14 +93,18 @@ export const RunTimeline = (props: Props) => {
...
 
-  const buckets = jobs.reduce(
-    (accum, job) => {
-      const {repoAddress} = job;
-      const repoKey = repoAddressAsURLString(repoAddress);
-      const jobsForRepo = accum[repoKey] || [];
-      return {...accum, [repoKey]: [...jobsForRepo, job]};
-    },
-    {} as Record<string, TimelineJob[]>,
+  const buckets = React.useMemo(
```
Comment (salazarm) [inline js_modules/dagster-ui/packages/ui-core/src/runs/RunTimeline.tsx:96]: This is really expensive so memoizing it

File: js_modules/dagster-ui/packages/ui-core/src/runs/useRunsForTimeline.tsx [hunk 2]
```diff
@@ -317,16 +327,71 @@ export const useRunsForTimeline = ({
...
+      jobs[jobKey] = {
+        key: jobKey,
+        jobName: isAdHoc ? 'Ad hoc materializations' : pipelineName,
+        jobType: isAdHoc ? 'asset' : 'job',
+        repoAddress,
+        path: workspacePipelinePath({
+          repoName: repoAddress.name,
+          repoLocation: repoAddress.location,
+          pipelineName,
+          isJob: true,
```
Comment (salazarm) [inline js_modules/dagster-ui/packages/ui-core/src/runs/useRunsForTimeline.tsx:378]: Pretty sure this is always true. I grepped for `resolve_isJob` in OSS and Internal, [this is the only implementation of it](https://github.com/dagster-io/dagster/blob/f6c842b29e514c0a5354e5dceabc4b7bc23a23e3/python_modules/dagster-graphql/dagster_graphql/schema/pipelines/pipeline.py#L904)

File: js_modules/dagster-ui/packages/ui-core/src/runs/RunTimeline.tsx (hunk 1 above)
Comment (alangenfeld) [reply js_modules/dagster-ui/packages/ui-core/src/runs/RunTimeline.tsx:96]: does this work as expected ? when i was poking around here all the object references were shifting

File: js_modules/dagster-ui/packages/ui-core/src/runs/fetchPaginatedBucketData.ts [hunk 3]
```diff
@@ -5,6 +5,11 @@ export type QueryResultData<DataType> = {
   called: boolean;
 };
 
+/**
+ * Don't use this function, its an awkward API really meant just for the RunTimeline
+ * setQueryData doesn't actually set any data, you have to do it yourself.
+ * In the case of the RunTimeline it doesn't set any data because it persists it to a cache instead
+ */
```
Comment (alangenfeld) [inline js_modules/dagster-ui/packages/ui-core/src/runs/fetchPaginatedBucketData.ts:12]: is there a better name for `setQueryData` then?

**Review by alangenfeld — APPROVED** [review]

File: js_modules/dagster-ui/packages/ui-core/src/runs/RunTimeline.tsx (hunk 1 above)
Comment (salazarm) [reply js_modules/dagster-ui/packages/ui-core/src/runs/RunTimeline.tsx:96]: The object references change but at least this memo doesn't refire if the component re-renders for other reasons

**Pull request merged**